### DISCUSSION TITLE
RateLimit support in PnP.Core?

### DISCUSSION THREAD

**Discussion by AndersRask** [pnp/pnpcore#1026] [Ideas] [answered]
Will we see support for RateLimit headers in PnP.Core?

We have already full support for Retry-After, so it would make sense to have a cool-down build into requests if RateLimit-Remaining drops below a certain percentage?

For context see Berts samples here
https://github.com/OneDrive/samples/tree/a875148366eddcd733e2b64e237b4a1343cfe299/scenarios/throttling-ratelimit-handling

There is a Ratelimits branch but it looks abandoned

**Comment by jansenbe** [comment, marked as answer]
@AndersRask : RateLimit support is integrated in PnP Core SDK (https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Services/Core/Http/RateLimiter.cs), but it's off by default. You can turn it on via the `RateLimiterMinimumCapacityLeft` configuration option (see https://pnp.github.io/pnpcore/using-the-sdk/basics-settings.html). I still need to beter document this feature, you can use the built-in functionality via the configuration option, but you can also opt to receive an event and handle the headers in your application. The Microsoft 365 Assessment tool uses that model as it combines PnP Core SDK with the CSOM libraries. See https://github.com/pnp/pnpassessment/blob/main/src/PnP.Scanning/PnP.Scanning.Core/Services/ScanManager.cs#L42-L76

**Reply by AndersRask** [reply]
Thanx for the swift reply @jansenbe 

Do you know if the guys at PnP.PowerShell have plans of adopting this in the PowerShell module?

**Reply by jansenbe** [reply]
@AndersRask : not sure, best to create an issue in the PS repo for that. PnP Framework also requires updates to fully support ratelimit header handling.

**Reply by AndersRask** [reply]
Right, the retryhandler is in framework too... I always seem to get Core and Framework confused :)
I will ask in Framework too then!

thanx

**Comment by carlosvigopaz** [comment]
Dear @jansenbe,

our client is part of a big corporation with many subsidiary companies and all of them use the same SharePoint tenant.
That is why he's worried that his provisioning solution based on **PowerShell scripts and PnP.PowerShell cmdlets** (v3.1) will get throttled.

So he has the following concerns:
- Is there a way to purposely reproduce the HTTP codes 429 or 503 in the responses sent back from SharePoint Online?
Because the HTTP packages that we've captured in our tests are always successful (200) and we'd like to be sure that PnP.PowerShell is handling the rate limit headers.

- Is there a way to manipulate/set rate limit values through the use of the PnP.PowerShell module, for example a settings file or a cdmlet we don't know?

Looking forward to your promptly response.

**Reply by jansenbe** [reply]
@carlosvigopaz : PnP PS authenticates via an Entra App and therefore is throttled based upon the resource limits described here: https://learn.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online. Using the Microsoft dev Proxy (https://learn.microsoft.com/en-us/microsoft-cloud/dev/dev-proxy/how-to/simulate-throttling-microsoft-365) you can emulate 429 and Rate-Limit responses to test PnP PS. 

I don't think there's a way in PnP PS to configure the Rate-Limiting behavior, best to ask to the PnP PS crew. For PnP Core SDK this can be done using https://pnp.github.io/pnpcore/using-the-sdk/basics-settings.html and the RateLimiterMinimumCapacityLeft property.

**Reply by carlosvigopaz** [reply]
@jansenbe thank you very much for your promptly response.
We'll try the Microsoft Dev Proxy.
